modelzoo/DBMTL/train.py

```diff
@@ -0,0 +1,709 @@
+import time
+import argparse
+import numbers
+import tensorflow as tf
+import os
+import sys
+import math
+import collections
+from tensorflow.python.client import timeline
+import json
+from tensorflow.python.ops import partitioned_variables
+
+if tf.__version__ >= '2.0':
+  tf = tf.compat.v1
+
+# Set to INFO for tracking training, default is WARN. ERROR for least messages
+tf.logging.set_verbosity(tf.logging.INFO)
+print("Using TensorFlow version %s" % (tf.__version__))
+SEED=2021
+'''
+INPUT CONFIG SPECIFICS
+'''
+TRAIN_DATA_NAME="taobao_train_data"
+TEST_DATA_NAME="taobao_test_data"
+
+LABEL_COLUMNS = ["clk", "buy"]
+HASH_INPUTS = [
+    "pid",
+    "adgroup_id",
+    "cate_id",
+    "campaign_id",
+    "customer",
+    "brand",
+    "user_id",
+    "cms_segid",
+    "cms_group_id",
+    "final_gender_code",
+    "age_level",
+    "pvalue_level",
+    "shopping_level",
+    "occupation",
+    "new_user_class_level",
+    "tag_category_list",
+    "tag_brand_list"
+    ]
+IDENTITY_INPUTS = ["price"]
+ALL_FEATURE_COLUMNS = HASH_INPUTS + IDENTITY_INPUTS
+ALL_INPUT = LABEL_COLUMNS + HASH_INPUTS + IDENTITY_INPUTS
+NOT_USED_CATEGORY = ["final_gender_code"]
+
+HASH_BUCKET_SIZES = {
+    'pid': 10,
+    'adgroup_id': 100000,
+    'cate_id': 10000,
+    'campaign_id': 100000,
+    'customer': 100000,
+    'brand': 100000,
+    'user_id': 100000,
+    'cms_segid': 100,
+    'cms_group_id': 100,
+    'final_gender_code': 10,
+    'age_level': 10,
+    'pvalue_level': 10,
+    'shopping_level': 10,
+    'occupation': 10,
+    'new_user_class_level': 10,
+    'tag_category_list': 100000,
+    'tag_brand_list': 100000
+}
+
+NUM_BUCKETS = {
+    'price': 50
+}
+
+def generate_taobao_input_data(filename, batch_size, num_epochs):
+    def parse_csv(value):
+        tf.logging.info('Parsing {}'.format(filename))
+        HASH_defaults = [[" "] for i in range(0, len(HASH_INPUTS))]
+        label_defaults = [[0] for i in range (0, len(LABEL_COLUMNS))]
+        IDENTITY_defaults = [[0] for i in range(0, len(IDENTITY_INPUTS))]
+        column_headers = LABEL_COLUMNS + HASH_INPUTS + IDENTITY_INPUTS
+        record_defaults = label_defaults + HASH_defaults + IDENTITY_defaults
+        columns = tf.io.decode_csv(value, record_defaults=record_defaults)
+        all_columns = collections.OrderedDict(zip(column_headers, columns))
+        labels = []
+        for i in range(0, len(LABEL_COLUMNS)):
+            labels.append(all_columns.pop(LABEL_COLUMNS[i]))
+        label = tf.stack(labels, axis=1)
+        features = all_columns
+        return features, label
+
+    # Extract lines from input files using the Dataset API.
+    dataset = tf.data.TextLineDataset(filename)
+    dataset = dataset.shuffle(buffer_size=400000,
+                              seed=SEED)  # set seed for reproducing
+    dataset = dataset.repeat(num_epochs)
+    dataset = dataset.batch(batch_size)
+    dataset = dataset.map(parse_csv, num_parallel_calls=28)
```

Comment:
I guess it might be better to change `28` to `tf.data.experimental.AUTOTUNE` but I'm not sure it will change the performance.